### PR TITLE
Rewriting QueryPlan

### PR DESCRIPTION
**Pull request by SvenLehmann**
This PR refactors the QueryPlan to handle deep nested queries and unify various functions.

### CODE REVIEW

File: src/test/scala/de/hpi/svedeb/ExternalUser.scala [hunk 1]
```diff
@@ -1,91 +1,84 @@
...
+    val result = block    // call-by-name
+    val t1 = System.nanoTime()
+    println(s"Elapsed time in $description: ${(t1 - t0)/1000000.0}ms")
+    result
+  }
 
-  val before = System.nanoTime()
-  testScan()
-  val after = System.nanoTime()
+  def runQuery(name: String, function: => Result): Unit = {
```
Comment (frederike-ramin) [inline src/test/scala/de/hpi/svedeb/ExternalUser.scala]: Smooth!

File: src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala [hunk 2]
```diff
@@ -1,50 +1,96 @@
...
   it should "update the assigned worker" in {
     val node = GetTable("S")
     val worker = TestProbe("createTableWorker")
-    node.updateAssignedWorker(worker.ref)
+    node.updateAssignedWorker(node, worker.ref)
 
     assert(node.assignedWorker == worker.ref)
   }
 
+  it should "find node with worker" in {
```
Comment (frederike-ramin) [inline src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala]: Empty test?

File: src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala [hunk 3]
```diff
@@ -56,21 +102,36 @@ class QueryPlanTest extends AbstractActorTest("QueryPlan") {
...
     val table = TestProbe("table")
 
-    firstNode.nextStage(worker.ref, table.ref, worker.ref, firstNode)
+    firstNode.advanceToNextStep(worker.ref, table.ref, worker.ref, firstNode)
+
     assert(firstNode.assignedWorker == worker.ref)
     assert(firstNode.resultTable == table.ref)
 
-    secondNode.nextStage(worker.ref, table.ref, worker.ref, firstNode)
+    secondNode.advanceToNextStep(worker.ref, table.ref, worker.ref, firstNode)
```
Comment (frederike-ramin) [inline src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala]: We're inconsistent in the usage of step and stage.

File: src/main/scala/de/hpi/svedeb/api/QueryPlanExecutor.scala [hunk 4]
```diff
@@ -12,81 +12,86 @@ object QueryPlanExecutor {
...
+                     worker: ActorRef,
+                     node: QueryPlanNode): APIWorkerState = {
+      val newQueryPlan = initialQueryPlan.findNodeAndUpdateWorker(worker, node)
+      APIWorkerState(sender, Some(newQueryPlan), queryId)
     }
 
-    def nextStage(lastWorker: ActorRef, resultTable: ActorRef, nextWorker: ActorRef, nextStep: QueryPlanNode): APIWorkerState = {
-      val newQueryPlan = queryPlan.nextStage(lastWorker, resultTable, nextWorker, nextStep)
-      APIWorkerState(newQueryPlan, sender, this.queryId)
+    def nextStage(initialQueryPlan: QueryPlanNode,
```
[the remark this answers is not in the document]
Comment (SvenLehmann) [reply src/main/scala/de/hpi/svedeb/api/QueryPlanExecutor.scala]: Agreed.

File: src/main/scala/de/hpi/svedeb/queryplan/QueryPlan.scala [hunk 5]
```diff
@@ -3,135 +3,131 @@ package de.hpi.svedeb.queryplan
...
-          else EmptyNode()
-        case DropTable(_) =>
-          if (this == node) this
-          else EmptyNode()
-        case InsertRow(table, _) =>
-          if (this == node) this
-          else table.findNode(node)
-        case EmptyNode() => this
+    def findNodeWithWorker(workerRef: ActorRef): Option[QueryPlanNode] = {
+      @tailrec
```
[the remark this answers is not in the document]
Comment (SvenLehmann) [reply src/main/scala/de/hpi/svedeb/queryplan/QueryPlan.scala]: Will do that, but tried to simplify as well.

File: src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala (hunk 3 above)
Comment (SvenLehmann) [reply src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala]: Changed all occurrences to 'stage' for now, but we can discuss the naming offline.

File: src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala (hunk 2 above)
Comment (SvenLehmann) [reply src/test/scala/de/hpi/svedeb/queryPlan/QueryPlanTest.scala]: Deleted it, this was already covered by another test.

File: src/test/scala/de/hpi/svedeb/ExternalUser.scala (hunk 1 above)
Comment (SvenLehmann) [reply src/test/scala/de/hpi/svedeb/ExternalUser.scala]: :)

**Pull request merged**